File: src/_41.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int len;
    int firstMissingPositive(vector<int> &nums)
    {
        len = nums.size();
        int ans = -1;
        for (int i = 0; i < len; ++i)
        {
            rectify(nums, i);
        }
        for (int i = 0; i < len; ++i)
        {
            if(nums[i] != i+1)
            {
                return i+1;
            }
        }
        return len+1;
    }
    inline void rectify(vector<int> &nums, int i)
    {
        int next;
        int cur = nums[i];
        while (cur <= len && cur > 0 && nums[cur-1] != cur)
        {
            next = nums[cur];
            nums[cur] = cur;
            cur = next;
        }
    }
};

int main()
{
    return 0;
}

File: 41.py
from typing import List

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        for i in range(n):
            if nums[i] <= 0:
                continue
            if nums[i] > n:
                nums[i] = -1
                continue
            if nums[i] == i + 1:
                continue
            t = nums[i]
            nums[i] = -1
            while t <= n and t > 0 and nums[t-1] != t:
                p = nums[t-1]
                nums[t-1] = t
                t = p
        for i, x in enumerate(nums):
            if x > n and x <= 0:
                return i + 1
        return n + 1
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        for i in range(n):
            while 1 <= nums[i] <= n and nums[i] != nums[nums[i]-1]:
                j = nums[i] - 1
                nums[i], nums[j] = nums[j], nums[i]
        for i in range(n):
            if nums[i] != i + 1:
                return i + 1
            
        return n + 1

t = Solution()
t.firstMissingPositive([2,1])